cruise-control/src/main/java/com/linkedin/kafka/cruisecontrol/analyzer/goals/BrokerSetAwareGoal.java

```diff
@@ -178,28 +182,38 @@ protected void updateGoalState(ClusterModel clusterModel, OptimizationOptions op
     // Sanity check: No replica should be moved to a broker, which used to host any replica of the same partition on its broken disk.
     GoalUtils.ensureReplicasMoveOffBrokersWithBadDisks(clusterModel, name());
     // Sanity check to confirm that the final distribution is broker set aware.
-    ensureBrokerSetAware(clusterModel, optimizationOptions);
+    ensureBrokerSetAware(clusterModel);
     if (_provisionResponse.status() != ProvisionStatus.OVER_PROVISIONED) {
       _provisionResponse = new ProvisionResponse(ProvisionStatus.RIGHT_SIZED);
     }
     finish();
   }
 
-  private void ensureBrokerSetAware(ClusterModel clusterModel, OptimizationOptions optimizationOptions)
+  private void ensureBrokerSetAware(ClusterModel clusterModel)
       throws OptimizationFailureException {
     // Sanity check to confirm that the final distribution is brokerSet aware.
     for (Map.Entry<String, List<Partition>> partitionsByTopic : clusterModel.getPartitionsByTopic().entrySet()) {
       String topicName = partitionsByTopic.getKey();
       if (!_excludedTopics.contains(topicName)) {
         List<Partition> partitions = partitionsByTopic.getValue();
-        Set<Broker> allBrokersForTopic = partitions.stream()
-                                                   .map(partition -> partition.partitionBrokers())
-                                                   .flatMap(brokers -> brokers.stream())
-                                                   .collect(Collectors.toSet());
-        // Checks if a topic's brokers do not all live in a single brokerSet
-        if (_brokersByBrokerSet.values().stream().noneMatch(brokerSetBrokers -> brokerSetBrokers.containsAll(allBrokersForTopic))) {
-          throw new OptimizationFailureException(
-              String.format("[%s] Topic %s is not brokerSet-aware. brokers (%s).", name(), topicName, allBrokersForTopic));
+        if (_isPartitionLevelBrokerSetAware) {
```

Comment:
Instead of doing this i would recommend introducing inheritance. 

1. Introduce a configuration - `broker.set.awareness.level=partition`
We can have only two choices - `partition` and `topic`
2. The Level can be picked from an ENUM so that no other level is allowed.
3. The default value of this configuration can be set to level `topic`.
4. Introduce an interface called `BrokerSetAwarenessValidator`
5. Implement two implementations `TopicLevelBrokerSetAwarenessValidator` and `PartitionLevelBrokerSetAwarenessValidator`
6. Based on configuration set, the instance of the BrokerSetAwarenessValidator will be chosen.
7. From here just call `brokerSetAwarenessValidator.validate(partitions)`

Reply:
You could also just introduce the validate() method in th ENUMs (with values PARTITION and TOPIC )

```
    private enum BrokerSetAwarenessLevel {
        TOPIC {
            public void validate(Partitions p) {
                // do something
            }
        },
        PARTITION {
            public void validate(Partitions p) {
                // do something different
            }
        }
```